common/src/autogluon/common/loaders/load_s3.py

```diff
@@ -27,47 +29,85 @@ def download(input_bucket, input_prefix, local_path):
 
 
-def list_bucket_prefix_s3(bucket, prefix):
-    return list_bucket_prefix_suffix_s3(bucket=bucket, prefix=prefix)
-
-
-def list_bucket_prefix_suffix_s3(bucket, prefix, suffix=None, banned_suffixes=None):
+# TODO: Expand to support arbitrary list of candidate file paths
+# TODO: add local file path support
+# TODO: Consider renaming and deprecating old name
+def list_bucket_prefix_suffix_contains_s3(bucket: str,
+                                          prefix: str,
+                                          suffix: Union[str, List[str]] = None,
+                                          banned_suffixes: List[str] = None,
+                                          contains: Union[str, List[str]] = None,
+                                          banned_contains: List[str] = None) -> List[str]:
+    """
+    Returns a list of file paths within an S3 bucket that satisfies the constraints.
+
+    Parameters
+    ----------
+    bucket : str
+        The S3 bucket to list files from.
+        You must have read permissions to the S3 bucket and its files for this function to work.
+    prefix : str
+        The string prefix to search for files within the S3 bucket. Any file outside of this prefix will not be considered.
+        For example, if `bucket='autogluon'` and `prefix='datasets/'`,
+        only files starting under `s3://autogluon/datasets/` will be considered.
+    suffix : str or List[str], default = None
+        If specified, filters files to ensure their paths end with the specified suffix (if str)
+        or at least one element of `suffix` (if list) in the post-prefix string path.
+    banned_suffixes : List[str], default = None
+        If specified, filters files to ensure their paths do not end with any element in `banned_suffixes`.
+    contains : str or List[str], default = None
+        If specified, will filter any result that doesn't contain `contains` (if str)
+        or at least one element of `contains` (if list) in the post-prefix string path.
+    banned_contains : List[str], default = None
+        If specified, filters files to ensure their paths do not contain any element in `banned_contains`.
+
+    Returns
+    -------
+    Returns a list of file paths within an S3 bucket that satisfies the constraints.
+
+    """
     import boto3
     if banned_suffixes is None:
         banned_suffixes = []
-    s3 = boto3.resource('s3')
-    my_bucket = s3.Bucket(bucket)
-    prefix = prefix
-
-    files = []
-    for object_summary in my_bucket.objects.filter(Prefix=prefix):
-        suffix_full = object_summary.key.split(prefix, 1)[1]
-        is_banned = False
-        for banned_suffix in banned_suffixes:
-            if banned_suffix in suffix_full:
-                is_banned = True
-        if (not is_banned) and ((suffix is None) or (suffix in suffix_full)):
-            files.append(object_summary.key)
-
-    return files
-
+    if banned_contains is None:
+        banned_contains = []
```

Comment:
1) consider using `exclude` vs `banned` - most unix tools use the first naming convention
2) consider using a single parameter supporting wildcards or regex - this will solve all possible use cases for filtering
`list_bucket_prefix_suffix_contains_s3(..., exclude=['**/*.bak', '**/data/*_excl.csv'])`

Reply:
1. Updated to `exclude`
2. Added TODO